### ISSUE TITLE
Language mixing (Swahili/Indonesian) and initialization latency with Gemma LiteRT (Android)

### ISSUE THREAD

**Issue by ishizuki-tech** [ishizuki-tech/SurveyNav2025#7]
# [MediaPipe Issue #3] Language mixing (Swahili/Indonesian) and initialization latency with Gemma LiteRT on Android

## Summary

Using `gemma-3n-E4B-it-int4.litertlm` with MediaPipe Tasks GenAI on Android, we observe:

1. **Language mixing:** Output often contains Indonesian (Bahasa Indonesia) despite Swahili-only instructions.
2. **Initialization latency:** `LlmInference.createFromOptions(...)` takes ~12 seconds on-device (Pixel 9a).

We are looking for guidance on:

* language control mechanisms (API knobs / best prompt patterns),
* expected/typical initialization performance for this model,
* recommended engine/session reuse and profiling practices,
* known looping behaviors (repeated `<end_of_turn>`).

## Environment

* **Platform:** Android (Kotlin)
* **Gradle config:** `minSdk=26`, `targetSdk=36`, `compileSdk=36`
* **Device:** Pixel 9a
* **Dependency (Gradle):**

  * `implementation(libs.mediapipe.tasks.genai)` // `com.google.mediapipe:tasks-genai:0.10.29`
* **Model:** `gemma-3n-E4B-it-int4.litertlm` (LiteRT format)
* **Delegate:** GPU (with CPU fallback)

Related debug notes: [https://github.com/ishizuki-tech/SurveyNav2025/issues/3](https://github.com/ishizuki-tech/SurveyNav2025/issues/3)

## Context

Use case: Swahili-language agricultural survey evaluation for Tanzania (farmer answers, follow-up questions, scoring).

We expect Swahili-only strings in structured responses (often JSON), but language mixing breaks UX and validation.

---

### Observed behavior

Even with explicit Swahili-only instructions, the model frequently outputs Indonesian words/phrases, sometimes mixed in a single response.

Examples observed:

* `Jawaban tidak lengkap` (Indonesian)
* `kerugian hasil tanaman akibat FAW selama tiga musim terakhir` (Indonesian)
* Mixed Swahili + Indonesian in one JSON field.

### Minimal reproduction sketch (Kotlin)

```kotlin
// NOTE: English-only comments by preference.

val engine = LlmInference.createFromOptions(
  context,
  LlmInference.LlmInferenceOptions.builder()
    .setModelPath(modelPath)
    .build()
)

val session = LlmInferenceSession.createFromOptions(
  engine,
  LlmInferenceSession.LlmInferenceSessionOptions.builder()
    .setTemperature(0.3f)
    .setTopK(10)
    .setMaxTokens(512)
    .build()
)

val userQuestion = "Unapoteza mavuno kiasi gani kutokana na funza wa jeshi vamizi (FAW) katika misimu mitatu iliyopita?"
val userAnswer = "Nimepoteza magunia 3 kwa ekari katika msimu uliopita."

val prompt = """
<start_of_turn>user
Wewe ni msaidizi wa tathmini ya majibu ya wakulima.
Tathmini jibu la mkulima kuhusu FAW (Fall Armyworm).

MUHIMU:
- Tumia Kiswahili PEKEE.
- USITUMIE Indonesian (mfano: "tidak", "adalah", "yang", "untuk", "dengan", "karena").
- Toa matokeo kwa sentensi fupi.

Swali: $userQuestion
Jibu: $userAnswer
<end_of_turn>
<start_of_turn>model
""".trim()

val sb = StringBuilder()
session.addQueryChunk(prompt)
session.generateResponseAsync { partial, done ->
  sb.append(partial)
  if (done) {
    val raw = sb.toString()
    // Inspect raw for Indonesian tokens.
  }
}
```

### Questions (Language)

1. Is Indonesian/Swahili mixing an expected behavior for `gemma-3n-E4B-it-int4.litertlm`?
2. Are there any MediaPipe API mechanisms to bias or constrain language generation?

   * language tags, locale settings, constraints, stop sequences, etc.
3. Are there recommended prompt patterns for better language adherence?

   * repeating the instruction, including Swahili-only exemplars, using language code "sw", etc.
4. Is there any way to get token-level diagnostics (token probabilities / logits / metadata) to understand why Indonesian tokens are chosen?

### Workarounds / best practices

We have tried:

* explicit “Swahili only” instructions (plus explicit Indonesian word blacklist)
* Swahili question text and Tanzanian farmer context

Questions:

1. Are there known workarounds to improve Swahili adherence with Gemma LiteRT models?
2. Should we consider a different model variant (size/quantization) for stronger Swahili behavior?
3. Is there any official guidance on language-specific fine-tuning for LiteRT models (or recommended alternatives)?

---

### Observed behavior

Engine creation latency:

* `LlmInference.createFromOptions(context, optionsWithGpuBackend)` takes **~12 seconds** on Pixel 9a.

Measurement details:

* Timer starts immediately before `createFromOptions(...)`
* Timer ends when the method returns (engine ready)
* This is measured on cold app start (first creation in process)

### Questions (Initialization)

1. Is ~12 seconds expected for this model on mobile?
2. What contributes most to initialization time?

   * model file IO, memory mapping/allocation, GPU kernel compilation/warmup, delegate setup, etc.
3. Recommended practices to reduce init time:

   * CPU vs GPU backend tradeoffs
   * smaller model / different quantization
   * preloading strategies (foreground service, splash warmup, caching)
   * keeping the engine resident vs re-creating

### Engine reuse patterns

Current pattern:

* Keep a single global `LlmInference` instance alive for app lifetime
* Reuse across requests

Questions:

1. Is this the recommended pattern for Android apps?
2. Memory implications:

   * does `LlmInference` retain the full model in RAM?
   * GPU memory usage expectations?
3. Any best practices for lifecycle:

   * unload/reload on background/foreground?
   * unload after inactivity window?

### Performance monitoring / profiling

Questions:

1. Any MediaPipe APIs/tools to report:

   * time-to-first-token, tokens/sec, per-token latency
   * memory usage (RAM / GPU)
   * GPU vs CPU split
2. Recommended profiling tools/workflow for Android + MediaPipe LLM:

   * Android Studio Profiler, Perfetto/Systrace, etc.
3. Realistic performance targets for this model class on mid-range devices:

   * tokens/sec, TTFT, latency for ~200 tokens

---

## Part C: Looping / watchdog interactions

We use a 20-second app-level watchdog that calls `cancelGenerateResponseAsync()` on timeout.

Observed issue:

* occasionally the model emits repeated `<end_of_turn>` tokens (loop-like behavior), requiring the watchdog.

Questions:

1. Are there known looping issues with Gemma LiteRT models for certain prompts/sampling settings?
2. Is looping more likely with low temperature / low topP, strict JSON requests, or explicit turn tokens?
3. Recommended sampling parameters to reduce loops while preserving output quality?
4. Recommended approach beyond an app-level watchdog?

---

## Impact

* Language mixing causes poor UX for Swahili-speaking users and reduces trust.
* Initialization latency (~12s) causes unacceptable app startup delay.
* Occasional loops require defensive cancellation/timeouts and complicate lifecycle management.

## Desired outcomes

1. Guidance on Swahili language control mechanisms (or confirmation of limitations)
2. Performance optimization recommendations for mobile
3. Best practices for engine reuse and lifecycle/memory management
4. Known limitations and recommended mitigations for looping behavior

Thank you for any insights!
